odoo/tests/common.py

```diff
@@ -309,64 +309,93 @@ def assertRecordValues(self, records, expected_values):
         :param expected_values:       List of dicts expected to be exactly matched in records
         '''
 
-        def _compare_candidate(record, candidate):
-            ''' Return True if the candidate matches the given record '''
-            for field_name in candidate.keys():
+        def _compare_candidate(record, candidate, field_names):
+            '''
+            Compare all the values in `candidate` with a record.
+
+            :param record:      record being compared
+            :param candidate:   dict of values to compare
+            :return:            a list of triplets representing the different values.
+                Triplet format: (field_name, record_value, candidate_value)
+
+                If this list is empty, it means that there is no difference between
+                `candidate` and `record`.
+            '''
+            diff = {}
+            for field_name in field_names:
                 record_value = record[field_name]
-                candidate_value = candidate[field_name]
                 field_type = record._fields[field_name].type
                 if field_type == 'monetary':
                     # Compare monetary field.
                     currency_field_name = record._fields[field_name].currency_field
                     record_currency = record[currency_field_name]
-                    if record_currency.compare_amounts(candidate_value, record_value)\
-                            if record_currency else candidate_value != record_value:
-                        return False
+                    if field_name not in candidate:
+                        diff[field_name] = (record_value, None)
+                    elif record_currency:
+                        if record_currency.compare_amounts(candidate[field_name], record_value):
+                            diff[field_name] = (record_value, record_currency.round(candidate[field_name]))
+                    elif candidate[field_name] != record_value:
+                        diff[field_name] = (record_value, candidate[field_name])
                 elif field_type in ('one2many', 'many2many'):
                     # Compare x2many relational fields.
                     # Empty comparison must be an empty list to be True.
-                    if set(record_value.ids) != set(candidate_value):
-                        return False
+                    if field_name not in candidate:
```

Comment:
This seems to start each of the sub-conditions; couldn't it be put just once, before "if field_type == 'monetary'" ?

Reply:
Not really because the difference to add is different depending the field type :/